<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to the refdb package • refdb</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to the refdb package">
<meta property="og:description" content="refdb">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">refdb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro_refdb.html">Introduction to the refdb package</a>
    </li>
    <li>
      <a href="../articles/ncbi_bold.html">Downloading and combining data from NCBI Genbank and BOLD</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fkeck/refdb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="intro_refdb_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="intro_refdb_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="intro_refdb_files/jquery-1.12.4/jquery.min.js"></script><link href="intro_refdb_files/leaflet-1.3.1/leaflet.css" rel="stylesheet">
<script src="intro_refdb_files/leaflet-1.3.1/leaflet.js"></script><link href="intro_refdb_files/leafletfix-1.0.0/leafletfix.css" rel="stylesheet">
<script src="intro_refdb_files/proj4-2.6.2/proj4.min.js"></script><script src="intro_refdb_files/Proj4Leaflet-1.0.1/proj4leaflet.js"></script><link href="intro_refdb_files/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet">
<script src="intro_refdb_files/leaflet-binding-2.0.4.1/leaflet.js"></script><link href="intro_refdb_files/leaflet-markercluster-1.0.5/MarkerCluster.css" rel="stylesheet">
<link href="intro_refdb_files/leaflet-markercluster-1.0.5/MarkerCluster.Default.css" rel="stylesheet">
<script src="intro_refdb_files/leaflet-markercluster-1.0.5/leaflet.markercluster.js"></script><script src="intro_refdb_files/leaflet-markercluster-1.0.5/leaflet.markercluster.freezable.js"></script><script src="intro_refdb_files/leaflet-markercluster-1.0.5/leaflet.markercluster.layersupport.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to the refdb package</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/fkeck/refdb/blob/master/vignettes/intro_refdb.Rmd"><code>vignettes/intro_refdb.Rmd</code></a></small>
      <div class="hidden name"><code>intro_refdb.Rmd</code></div>

    </div>

    
    
<p>Developing and maintaining a DNA reference database can be difficult. The <code>refdb</code> package has been developed to assist you during the different steps of the process. This vignette will present some of the important functions of the package and help you to get started with it.</p>
<div id="installing-and-loading-the-package" class="section level2">
<h2 class="hasAnchor">
<a href="#installing-and-loading-the-package" class="anchor"></a>Installing and loading the package</h2>
<p>The refdb package is available on CRAN which means that you can easily install the latest stable version using the RStudio IDE or with the function <code>install.packages</code>. In this tutorial we will also be using functions from several packages of the tidyverse. Therefore, if the <code>tidyverse</code> package is not already installed on your computer, you are encouraged to install it now.</p>
<p>Once everything is ready, we can start by loading the packages we need:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://fkeck.github.io/refdb/">refdb</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://fkeck.github.io/refdb/">refdb</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://forcats.tidyverse.org">forcats</a></span><span class="op">)</span></code></pre></div>
<div id="reading-data" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-data" class="anchor"></a>Reading data</h3>
<div id="reading-data-from-a-file" class="section level4">
<h4 class="hasAnchor">
<a href="#reading-data-from-a-file" class="anchor"></a>Reading data from a file</h4>
<p>In <code>refdb</code>, a reference database is a simple dataframe. This means that any functions able to read external data and return a dataframe in your R environment can be used to import a reference library. Here, for simplicity, we will assume the reference database is stored as a CSV on the disk. You can download this file <a href="https://raw.githubusercontent.com/fkeck/refdb/acee2ff1b3a9dca27aa7aeb9e0f1e5837105b094/inst/extdata/ephem.csv">here</a> (use “Save as”). Then, we can read the file from its location on your computer with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lib</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"my_path/ephem.csv"</span><span class="op">)</span></code></pre></div>
<p>Note that we use the function <code>read_csv</code> from tidyverse which returns a tibble. A tibble is a dataframe with some extra refinements. There is no issue with using tibbles with refdb functions (this is actually encouraged).</p>
</div>
<div id="setting-fields" class="section level4">
<h4 class="hasAnchor">
<a href="#setting-fields" class="anchor"></a>Setting fields</h4>
<p>Now we have a dataframe (a tibble) in our environment. However, the functions of the refdb package can do nothing from it because they need to know what is the actual content stored in each column. You have to provide this information, and this can be done using the function <code>refdb_set_fields</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lib</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_set_fields.html">refdb_set_fields</a></span><span class="op">(</span><span class="va">lib</span>,
                        taxonomy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>family <span class="op">=</span> <span class="st">"family_name"</span>,
                                     genus <span class="op">=</span> <span class="st">"genus_name"</span>,
                                     species <span class="op">=</span> <span class="st">"species_name"</span><span class="op">)</span>,
                        sequence <span class="op">=</span> <span class="st">"DNA_seq"</span>,
                        marker <span class="op">=</span> <span class="st">"marker"</span><span class="op">)</span></code></pre></div>
<p>Here we explicitly set the taxonomy, sequence and marker fields by indicating which columns of the dataframe correspond to each of those fields.</p>
<p>If you print the object <code>lib</code> now, you will see that the nucleotide symbols in the columns DNA_seq are colored. This is because this column has been identified as containing DNA sequences. A simple way to see which fields have been associated with which column is simply to use the function <code>refdb_get_fields</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/refdb_get_fields.html">refdb_get_fields</a></span><span class="op">(</span><span class="va">lib</span><span class="op">)</span>
<span class="co">#&gt; taxonomy:</span>
<span class="co">#&gt;   family: family_name</span>
<span class="co">#&gt;   genus: genus_name</span>
<span class="co">#&gt;   species: species_name</span>
<span class="co">#&gt; sequence: DNA_seq</span>
<span class="co">#&gt; marker: marker</span></code></pre></div>
<p>Note that if you are importing data from some popular reference databases, these functions can save you time: <code>refdb_set_fields_BOLD</code>, <code>refdb_set_fields_NCBI</code>, <code>refdb_set_fields_PR2</code>, <code>refdb_set_fields_diatbarcode</code>.</p>
</div>
</div>
<div id="cleaning-taxonomic-names-and-dna-sequences" class="section level3">
<h3 class="hasAnchor">
<a href="#cleaning-taxonomic-names-and-dna-sequences" class="anchor"></a>Cleaning taxonomic names and DNA sequences</h3>
<p>Now that the data are loaded and the fields have been set, we can start to work with the reference database. The first thing we will look at, is how to clean taxonomic names and DNA sequences. Cleaning is different than filtering (that we will see in the next section) in that the objective here is not to remove entries but to improve them.</p>
<p>Both sequences and taxonomic names can be affected by countless problems that would be impossible to list here. However, the refdb package provides functions to deal with many of the common cases. Let’s review some of them.</p>
<div id="missing-taxonomic-information" class="section level4">
<h4 class="hasAnchor">
<a href="#missing-taxonomic-information" class="anchor"></a>Missing taxonomic information</h4>
<p>Let’s take a look at the taxonomy of the 91st record:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lib</span> <span class="op">[</span><span class="fl">91</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 1 x 4</span>
<span class="co">#&gt;   order_name    family_name     genus_name species_name  </span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;           &lt;chr&gt;      &lt;chr&gt;         </span>
<span class="co">#&gt; 1 Ephemeroptera Leptophlebiidae Thraulodes Thraulodes sp.</span></code></pre></div>
<p>Here, the specimen has been identified at the genus level and the species is reported using the notation “sp.”. This is a common notation but not really a practical one when you manage a database. How would you report the family of an organism identified at the class level? In R, a general and efficient way to code for a missing information is simply to use the built-in <code>NA</code> indicator.</p>
<p>The function <code>refdb_clean_tax_NA</code> can be used to identify taxonomic levels for which there are no information and change them to <code>NA</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lib_na</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_clean_tax_NA.html">refdb_clean_tax_NA</a></span><span class="op">(</span><span class="va">lib</span><span class="op">)</span>
<span class="va">lib_na</span> <span class="op">[</span><span class="fl">91</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 1 x 4</span>
<span class="co">#&gt;   order_name    family_name     genus_name species_name</span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;           &lt;chr&gt;      &lt;chr&gt;       </span>
<span class="co">#&gt; 1 Ephemeroptera Leptophlebiidae Thraulodes &lt;NA&gt;</span></code></pre></div>
</div>
<div id="extra-information-appended-to-taxonomic-names" class="section level4">
<h4 class="hasAnchor">
<a href="#extra-information-appended-to-taxonomic-names" class="anchor"></a>Extra information appended to taxonomic names</h4>
<p>It is common to find taxonomic names to which various information have been added. See the 32nd row of our database:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lib</span> <span class="op">[</span><span class="fl">32</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 1 x 4</span>
<span class="co">#&gt;   order_name    family_name   genus_name species_name           </span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;         &lt;chr&gt;      &lt;chr&gt;                  </span>
<span class="co">#&gt; 1 Ephemeroptera Heptageniidae Epeorus    Epeorus longimanus B BG</span></code></pre></div>
<p>The species name is tagged “B BG” which probably meant something to the person who recorded this entry but is not useful to us and may even be problematic. The function <code>refdb_clean_tax_remove_extra</code> uses some simple rules to detect these information and remove them.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lib_na_extra</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_clean_tax_remove_extra.html">refdb_clean_tax_remove_extra</a></span><span class="op">(</span><span class="va">lib_na</span><span class="op">)</span>
<span class="va">lib_na_extra</span> <span class="op">[</span><span class="fl">32</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 1 x 4</span>
<span class="co">#&gt;   order_name    family_name   genus_name species_name      </span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;         &lt;chr&gt;      &lt;chr&gt;             </span>
<span class="co">#&gt; 1 Ephemeroptera Heptageniidae Epeorus    Epeorus longimanus</span></code></pre></div>
</div>
<div id="a-note-on-false-positive-and-negative" class="section level4">
<h4 class="hasAnchor">
<a href="#a-note-on-false-positive-and-negative" class="anchor"></a>A note on false positive and negative</h4>
<p>The cleaning functions of refdb uses several rules and heuristics, some of them can produce false positive and negative. You should always be careful and if you find cases where a function does not what it is expected to do, please report <a href="https://github.com/fkeck/refdb/issues">here</a>.</p>
</div>
<div id="cleaning-dna-sequences" class="section level4">
<h4 class="hasAnchor">
<a href="#cleaning-dna-sequences" class="anchor"></a>Cleaning DNA sequences</h4>
<p>Similarly it is possible to clean the sequences with different functions. For example, if we want to remove the gaps in the sequences:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">lib</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 8</span>
<span class="co">#&gt;   order_name  family_name genus_name species_name marker DNA_seq      lat    lon</span>
<span class="co">#&gt;   &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;  &lt;DNA&gt;      &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt; 1 Ephemeropt… Heptagenii… Epeorus    Epeorus fra… COI-5P ATAGTCGGA…  39.5  -79.2</span>
<span class="co">#&gt; 2 Ephemeropt… Oligoneuri… Oligoneur… Oligoneurie… COI-5P GACTTTATA…  42.4   22.5</span>
<span class="co">#&gt; 3 Ephemeropt… Ephemerell… Serratella Serratella … COI-5P ---------…  34.2 -118. </span>
<span class="co">#&gt; 4 Ephemeropt… Baetidae    Fallceon   &lt;NA&gt;         COI-5P ---------…  34.2 -119. </span>
<span class="co">#&gt; 5 Ephemeropt… Heptagenii… Heptagenia Heptagenia … COI-5P AACTTTATA…  58.7  -94.2</span>
<span class="co">#&gt; 6 Ephemeropt… Heptagenii… Epeorus    Epeorus vit… COI-5P AACTCTATA…  46.8  -66.1</span>

<span class="va">lib_nogap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_clean_seq_remove_gaps.html">refdb_clean_seq_remove_gaps</a></span><span class="op">(</span><span class="va">lib</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">lib_nogap</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 8</span>
<span class="co">#&gt;   order_name  family_name genus_name species_name marker DNA_seq      lat    lon</span>
<span class="co">#&gt;   &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;  &lt;DNA&gt;      &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt; 1 Ephemeropt… Heptagenii… Epeorus    Epeorus fra… COI-5P ATAGTCGGA…  39.5  -79.2</span>
<span class="co">#&gt; 2 Ephemeropt… Oligoneuri… Oligoneur… Oligoneurie… COI-5P GACTTTATA…  42.4   22.5</span>
<span class="co">#&gt; 3 Ephemeropt… Ephemerell… Serratella Serratella … COI-5P TGAGTTTTT…  34.2 -118. </span>
<span class="co">#&gt; 4 Ephemeropt… Baetidae    Fallceon   &lt;NA&gt;         COI-5P TAAGATTTT…  34.2 -119. </span>
<span class="co">#&gt; 5 Ephemeropt… Heptagenii… Heptagenia Heptagenia … COI-5P AACTTTATA…  58.7  -94.2</span>
<span class="co">#&gt; 6 Ephemeropt… Heptagenii… Epeorus    Epeorus vit… COI-5P AACTCTATA…  46.8  -66.1</span></code></pre></div>
</div>
<div id="other-cleaning-functions" class="section level4">
<h4 class="hasAnchor">
<a href="#other-cleaning-functions" class="anchor"></a>Other cleaning functions</h4>
<p>We have only seen a brief overview of the functions available for cleaning up the sequences and their associated taxonomy. It would be tedious to list all the functions and their options here. Instead, the table below summarises the functions currently available in refdb.</p>
<table class="table">
<colgroup>
<col width="40%">
<col width="60%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>refdb_clean_seq_remove_gaps</td>
<td>Remove gaps from genetic sequences</td>
</tr>
<tr class="even">
<td>refdb_clean_seq_remove_sideN</td>
<td>Remove repeated side N from genetic sequences</td>
</tr>
<tr class="odd">
<td>refdb_clean_seq_crop_primers</td>
<td>Crop genetic sequences with a set of primers</td>
</tr>
<tr class="even">
<td>refdb_clean_tax_remove_blank</td>
<td>Remove blank characters from taxonomic names</td>
</tr>
<tr class="odd">
<td>refdb_clean_tax_remove_extra</td>
<td>Remove extra words from taxonomic names</td>
</tr>
<tr class="even">
<td>refdb_clean_tax_harmonize_nomenclature</td>
<td>Harmonize taxonomic name nomenclature</td>
</tr>
<tr class="odd">
<td>refdb_clean_tax_remove_uncertainty</td>
<td>Remove terms indicating uncertainty in taxonomic names</td>
</tr>
<tr class="even">
<td>refdb_clean_tax_remove_subsp</td>
<td>Remove subspecific information from taxonomic names</td>
</tr>
<tr class="odd">
<td>refdb_clean_tax_NA</td>
<td>Convert missing taxonomic names to NA</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="filtering-records" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-records" class="anchor"></a>Filtering records</h3>
<p>Filtering is the step that usually follows cleaning and whose objective is to identify and remove records (rows) which do not satisfy quality criteria (those criteria and associated thresholds are of course specific to each project).</p>
<p>For example we may want to exclude very short sequences whose length is likely shorter than the targeted genetic marker. Let’s take a look at the distribution of sequence lengths in the database:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/refdb_plot_seqlen_hist.html">refdb_plot_seqlen_hist</a></span><span class="op">(</span><span class="va">lib</span><span class="op">)</span>
<span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></code></pre></div>
<p><img src="intro_refdb_files/figure-html/unnamed-chunk-12-1.png" width="432"></p>
<p>To exclude all the sequences shorter than 500 nucleotides we use the function <code>refdb_filter_seq_length</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lib_long</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_filter_seq_length.html">refdb_filter_seq_length</a></span><span class="op">(</span><span class="va">lib</span>, min_len <span class="op">=</span> <span class="fl">500</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">lib</span><span class="op">)</span>
<span class="co">#&gt; [1] 100</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">lib_long</span><span class="op">)</span>
<span class="co">#&gt; [1] 94</span></code></pre></div>
<p>The function has removed 6 records.</p>
<p>Another, more sophisticated filtering function is <code><a href="../reference/refdb_filter_seq_primer.html">refdb_filter_seq_primer()</a></code> which can find sequences which do not contain a motif (typically a primer sequence) and remove them from the reference database</p>
<p>For example, a commonly used forward primer for metazoan is the mlCOIintF (Leray 2013). I we want to keep only the sequences where this primer is found we can use:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/refdb_filter_seq_primer.html">refdb_filter_seq_primer</a></span><span class="op">(</span><span class="va">lib</span>, primer_forward <span class="op">=</span> <span class="st">"GGWACWGGWTGAACWGTWTAYCCYCC"</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 76 x 8</span>
<span class="co">#&gt;    order_name  family_name genus_name species_name marker DNA_seq     lat    lon</span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;  &lt;DNA&gt;     &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1 Ephemeropt… Oligoneuri… Oligoneur… Oligoneurie… COI-5P GACTTTAT…  42.4   22.5</span>
<span class="co">#&gt;  2 Ephemeropt… Ephemerell… Serratella Serratella … COI-5P --------…  34.2 -118. </span>
<span class="co">#&gt;  3 Ephemeropt… Baetidae    Fallceon   &lt;NA&gt;         COI-5P --------…  34.2 -119. </span>
<span class="co">#&gt;  4 Ephemeropt… Heptagenii… Heptagenia Heptagenia … COI-5P AACTTTAT…  58.7  -94.2</span>
<span class="co">#&gt;  5 Ephemeropt… Heptagenii… Epeorus    Epeorus vit… COI-5P AACTCTAT…  46.8  -66.1</span>
<span class="co">#&gt;  6 Ephemeropt… Ephemerell… Ephemerel… Ephemerella… COI-5P AACTTTAT…  46.0  -66.7</span>
<span class="co">#&gt;  7 Ephemeropt… Baetidae    Baetis     Baetis huds… COI-5P CGTTATAT…  67.8 -115. </span>
<span class="co">#&gt;  8 Ephemeropt… Polymitarc… Ephoron    Ephoron shi… COI-5P GAACTTCT…  34.1  134. </span>
<span class="co">#&gt;  9 Ephemeropt… Ephemerell… Serratella Serratella … COI-5P AACTTTAT…  34.2 -118. </span>
<span class="co">#&gt; 10 Ephemeropt… Ephemerell… Ephemerel… Ephemerella… COI-5P ACTTTATA…  39.9  -75.8</span>
<span class="co">#&gt; # … with 66 more rows</span></code></pre></div>
<p>You can see that only 76 sequences have passed the filter. By default the tolerance of error is 10%. We can change this value using the <code>max_error_forward</code> argument.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/refdb_filter_seq_primer.html">refdb_filter_seq_primer</a></span><span class="op">(</span><span class="va">lib</span>, primer_forward <span class="op">=</span> <span class="st">"GGWACWGGWTGAACWGTWTAYCCYCC"</span>, max_error_forward <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 99 x 8</span>
<span class="co">#&gt;    order_name  family_name genus_name species_name marker DNA_seq    lat     lon</span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;  &lt;DNA&gt;    &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1 Ephemeropt… Heptagenii… Epeorus    Epeorus fra… COI-5P ATAGTCG…  39.5  -79.2 </span>
<span class="co">#&gt;  2 Ephemeropt… Oligoneuri… Oligoneur… Oligoneurie… COI-5P GACTTTA…  42.4   22.5 </span>
<span class="co">#&gt;  3 Ephemeropt… Ephemerell… Serratella Serratella … COI-5P -------…  34.2 -118.  </span>
<span class="co">#&gt;  4 Ephemeropt… Baetidae    Fallceon   &lt;NA&gt;         COI-5P -------…  34.2 -119.  </span>
<span class="co">#&gt;  5 Ephemeropt… Heptagenii… Heptagenia Heptagenia … COI-5P AACTTTA…  58.7  -94.2 </span>
<span class="co">#&gt;  6 Ephemeropt… Heptagenii… Epeorus    Epeorus vit… COI-5P AACTCTA…  46.8  -66.1 </span>
<span class="co">#&gt;  7 Ephemeropt… Ephemerell… Ephemerel… Ephemerella… COI-5P AACTTTA…  46.0  -66.7 </span>
<span class="co">#&gt;  8 Ephemeropt… Baetidae    Baetis     Baetis huds… COI-5P CGTTATA…  67.8 -115.  </span>
<span class="co">#&gt;  9 Ephemeropt… Polymitarc… Ephoron    Ephoron shi… COI-5P GAACTTC…  34.1  134.  </span>
<span class="co">#&gt; 10 Ephemeropt… Baetidae    Baetis     Baetis rhod… COI-5P AGCTCGG…  52.2   -3.80</span>
<span class="co">#&gt; # … with 89 more rows</span></code></pre></div>
<p>By increasing the maximum error rate to 20%, we now retain 99 sequences.</p>
<p>Similarly to the the cleaning functions, there are many filtering functions available in refdb, as shown in the table below:</p>
<table class="table">
<colgroup>
<col width="40%">
<col width="60%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>refdb_filter_seq_length</td>
<td>Filter sequences based on their number of characters.</td>
</tr>
<tr class="even">
<td>refdb_filter_seq_ambiguous</td>
<td>Filter sequences based on their number of ambiguous characters.</td>
</tr>
<tr class="odd">
<td>refdb_filter_seq_homopolymers</td>
<td>Filter sequences based on their number of repeated characters.</td>
</tr>
<tr class="even">
<td>refdb_filter_seq_duplicates</td>
<td>Filter duplicated sequences.</td>
</tr>
<tr class="odd">
<td>refdb_filter_seq_stopcodon</td>
<td>Filter sequences based on their number of stop codons.</td>
</tr>
<tr class="even">
<td>refdb_filter_seq_primer</td>
<td>Filter sequences based on the presence of primers.</td>
</tr>
<tr class="odd">
<td>refdb_filter_tax_precision</td>
<td>Filter records based on their taxonomic precision.</td>
</tr>
<tr class="even">
<td>refdb_filter_ref_scope</td>
<td>Filter records by taxonomic scope of studies.</td>
</tr>
</tbody>
</table>
</div>
<div id="visualizing-the-reference-library" class="section level3">
<h3 class="hasAnchor">
<a href="#visualizing-the-reference-library" class="anchor"></a>Visualizing the reference library</h3>
<p>Now let’s take a tour of the functions you can use to produce graphical representation of your reference database. The fact that refdb stores reference database as dataframes makes it really easy to produce plots (e.g. with the tidyverse and <code>ggplot2</code>). For example, we can make a barplot showing the distribution of the families like this:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lib</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">family_name</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html">fct_reorder</a></span><span class="op">(</span><span class="va">family_name</span>, <span class="va">n</span>, .desc <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Number of records"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="intro_refdb_files/figure-html/unnamed-chunk-16-1.png" width="672"></p>
<p>Additionally, refdb provides ready-to use functions to produce more sophisticated plots. For example to represent the taxonomic coverage of your reference library across multiple levels with a tree map.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/refdb_plot_tax_treemap.html">refdb_plot_tax_treemap</a></span><span class="op">(</span><span class="va">lib</span><span class="op">)</span></code></pre></div>
<p><img src="intro_refdb_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<p>Alternatively you can represent this information with a taxonomic tree. These functions have several parameters to control what is represented in the plot.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/refdb_plot_tax_tree.html">refdb_plot_tax_tree</a></span><span class="op">(</span><span class="va">lib</span><span class="op">)</span></code></pre></div>
<p><img src="intro_refdb_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
<p>We conclude our overview of the graphical features of refdb with the <code>refdb_plot_map</code> function which can be used to generate interactive maps from the data. Of course this requires the sequences to be geographically located. As we have not yet associated columns to the latitude and longitude fields, we have to do this beforehand.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lib</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_set_fields.html">refdb_set_fields</a></span><span class="op">(</span><span class="va">lib</span>, latitude <span class="op">=</span> <span class="st">"lat"</span>, longitude <span class="op">=</span> <span class="st">"lon"</span><span class="op">)</span>
<span class="fu"><a href="../reference/refdb_plot_map.html">refdb_plot_map</a></span><span class="op">(</span><span class="va">lib</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-7901b694a7c3cbe5c937" style="width:672px;height:480px;" class="leaflet html-widget"></div>
<script type="application/json" data-for="htmlwidget-7901b694a7c3cbe5c937">{"x":{"options":{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}}},"calls":[{"method":"addTiles","args":["//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",null,null,{"minZoom":0,"maxZoom":18,"tileSize":256,"subdomains":"abc","errorTileUrl":"","tms":false,"noWrap":false,"zoomOffset":0,"zoomReverse":false,"opacity":1,"zIndex":1,"detectRetina":false,"attribution":"&copy; <a href=\"http://openstreetmap.org\">OpenStreetMap<\/a> contributors, <a href=\"http://creativecommons.org/licenses/by-sa/2.0/\">CC-BY-SA<\/a>"}]},{"method":"addCircleMarkers","args":[[39.541,42.4275,34.241,34.201,58.6756,46.84,45.976,67.809,34.13,52.1931,9.97,34.237,39.8596,49.023,39.021,38.564,34.23,46.1798,40.071,-39.12,8.639,41.9672,42.3118,39.44,38.499,39.169,44.296,34.649,46.807,54.642,58.663,40.919,50.8,39.61,39.481,34.274,34.205,58.761,41.606,46.953,-0.44924,40.375,42.368,9.0796,63.405,50.659,8.77,39.558,-33.147,42.534,36.139,58.6632,58.76,46.971,40.155,40.7,48.51,47.637,49.318,9.94,47.2354,39.049,39.126,41.8287,36.966,45.9151,-0.25,69,44.1037,-24.3067,-21.8631,34.88,42.8859,35.634,52.358,45.068,39.687,-28.447,44.6972,40.517,34.9389,0.226,44.387,44.6286,52.086,39.1174,54.446,46.1264,40.037,-0.244,8.667,33.7894,34.229,35.467,45.1509,38.436,-36.953,42.155,-42.15,50.662],[-79.2142,22.5202,-117.869,-119.001,-94.1688,-66.077,-66.719,-115.199,134.47,-3.7999,-84.082,-117.817,-75.7843,-114.045,-76.0472,-76.7601,-117.78,-66.6205,-105.615,174.51,-83.467,-82.5038,-81.853,-76.698,-76.903,-76.972,-78.452,45.387,-60.732,-122.742,-94.167,-105.498,7.32,-77.416,-76.698,-118.315,-118.166,-93.952,-76.049,-65.871,-77.943,-105.227,11.989,77.2319,10.383,9.629,77.41,-78.9084,151.258,23.5253,-79.17,-94.1679,-93.9521,-65.853,-105.466,-105.441,10.39,11.594,17.639,-84.073,89.7021,-76.095,-75.975,24.4476,44.392,-66.6169,-77.935,20.53,11.1136,132.905,121.008,134.93,13.2794,-83.084,-107.181,9.646,-78.456,152.998,-78.7431,-105.659,135.86,34.858,-73.825,-78.8614,-102.814,-75.9572,-122.653,-66.6116,-105.419,-77.985,-83.509,-81.9297,-117.768,-83.8784,-81.4702,-77.2524,148.325,23.4002,171.82,9.624],8,null,null,{"interactive":true,"className":"","stroke":false,"color":"#03F","weight":5,"opacity":0.5,"fill":true,"fillColor":"red","fillOpacity":1},{"showCoverageOnHover":true,"zoomToBoundsOnClick":true,"spiderfyOnMaxZoom":true,"removeOutsideVisibleBounds":true,"spiderLegPolylineOptions":{"weight":1.5,"color":"#222","opacity":0.5},"freezeAtZoom":false},null,["<h3>ID: 1<\/h3><b>Species_name: <\/b>Epeorus fragilis<br><b>Sequence length: <\/b>591","<h3>ID: 2<\/h3><b>Species_name: <\/b>Oligoneuriella rhenana<br><b>Sequence length: <\/b>658","<h3>ID: 3<\/h3><b>Species_name: <\/b>Serratella micheneri<br><b>Sequence length: <\/b>658","<h3>ID: 4<\/h3><b>Genus_name: <\/b>Fallceon<br><b>Sequence length: <\/b>658","<h3>ID: 5<\/h3><b>Species_name: <\/b>Heptagenia pulla<br><b>Sequence length: <\/b>658","<h3>ID: 6<\/h3><b>Species_name: <\/b>Epeorus vitreus<br><b>Sequence length: <\/b>658","<h3>ID: 7<\/h3><b>Species_name: <\/b>Ephemerella subvaria<br><b>Sequence length: <\/b>658","<h3>ID: 8<\/h3><b>Species_name: <\/b>Baetis hudsonicus<br><b>Sequence length: <\/b>654","<h3>ID: 9<\/h3><b>Species_name: <\/b>Ephoron shigae<br><b>Sequence length: <\/b>553","<h3>ID: 10<\/h3><b>Species_name: <\/b>Baetis rhodani<br><b>Sequence length: <\/b>572","<h3>ID: 11<\/h3><b>Species_name: <\/b>Leptohyphes SADR03<br><b>Sequence length: <\/b>658","<h3>ID: 12<\/h3><b>Species_name: <\/b>Serratella micheneri<br><b>Sequence length: <\/b>658","<h3>ID: 13<\/h3><b>Species_name: <\/b>Ephemerella dorothea<br><b>Sequence length: <\/b>657","<h3>ID: 14<\/h3><b>Species_name: <\/b>Callibaetis ferrugineus<br><b>Sequence length: <\/b>658","<h3>ID: 15<\/h3><b>Species_name: <\/b>Maccaffertium smithae<br><b>Sequence length: <\/b>604","<h3>ID: 16<\/h3><b>Species_name: <\/b>Ephemerella invaria<br><b>Sequence length: <\/b>588","<h3>ID: 17<\/h3><b>Species_name: <\/b>Baetis tricaudatus<br><b>Sequence length: <\/b>658","<h3>ID: 18<\/h3><b>Genus_name: <\/b>Cinygmula<br><b>Sequence length: <\/b>658","<h3>ID: 19<\/h3><b>Species_name: <\/b>Isothraulus abditus<br><b>Sequence length: <\/b>658","<h3>ID: 20<\/h3><b>Species_name: <\/b>Vacupernius packeri<br><b>Sequence length: <\/b>658","<h3>ID: 21<\/h3><b>Genus_name: <\/b>Caenis<br><b>Sequence length: <\/b>658","<h3>ID: 22<\/h3><b>Genus_name: <\/b>Caenis<br><b>Sequence length: <\/b>622","<h3>ID: 23<\/h3><b>Species_name: <\/b>Ephemerella invaria<br><b>Sequence length: <\/b>658","<h3>ID: 24<\/h3><b>Species_name: <\/b>Acerpenna macdunnoughi<br><b>Sequence length: <\/b>613","<h3>ID: 25<\/h3><b>Species_name: <\/b>Ephemerella dorothea<br><b>Sequence length: <\/b>463","<h3>ID: 26<\/h3><b>Species_name: <\/b>Callibaetis ferrugineus<br><b>Sequence length: <\/b>574","<h3>ID: 27<\/h3><b>Species_name: <\/b>Ephemerella dorothea<br><b>Sequence length: <\/b>661","<h3>ID: 28<\/h3><b>Species_name: <\/b>Tricorythodes mosegus<br><b>Sequence length: <\/b>658","<h3>ID: 29<\/h3><b>Species_name: <\/b>Baetis phoebus<br><b>Sequence length: <\/b>658","<h3>ID: 30<\/h3><b>Species_name: <\/b>Epeorus longimanus B BG<br><b>Sequence length: <\/b>606","<h3>ID: 31<\/h3><b>Species_name: <\/b>Baetis rhodani<br><b>Sequence length: <\/b>658","<h3>ID: 32<\/h3><b>Species_name: <\/b>Maccaffertium vicarium<br><b>Sequence length: <\/b>658","<h3>ID: 33<\/h3><b>Species_name: <\/b>Ephemerella invaria<br><b>Sequence length: <\/b>658","<h3>ID: 34<\/h3><b>Species_name: <\/b>Baetis adonis<br><b>Sequence length: <\/b>624","<h3>ID: 35<\/h3><b>Species_name: <\/b>Baetis adonis<br><b>Sequence length: <\/b>422","<h3>ID: 36<\/h3><b>Species_name: <\/b>Baetis brunneicolor<br><b>Sequence length: <\/b>658","<h3>ID: 37<\/h3><b>Species_name: <\/b>Callibaetis fluctuans<br><b>Sequence length: <\/b>658","<h3>ID: 38<\/h3><b>Species_name: <\/b>Nixe rusticalis<br><b>Sequence length: <\/b>658","<h3>ID: 39<\/h3><b>Genus_name: <\/b>Baetodes<br><b>Sequence length: <\/b>658","<h3>ID: 40<\/h3><b>Species_name: <\/b>Baetis tricaudatus D BG<br><b>Sequence length: <\/b>658","<h3>ID: 41<\/h3><b>Species_name: <\/b>Heptagenia longicauda<br><b>Sequence length: <\/b>594","<h3>ID: 42<\/h3><b>Species_name: <\/b>Afronurus kumbakkaraiensis<br><b>Sequence length: <\/b>658","<h3>ID: 43<\/h3><b>Species_name: <\/b>Heptagenia sulphurea<br><b>Sequence length: <\/b>611","<h3>ID: 44<\/h3><b>Species_name: <\/b>Baetis liebenauae<br><b>Sequence length: <\/b>640","<h3>ID: 45<\/h3><b>Species_name: <\/b>Cloeon bicolor<br><b>Sequence length: <\/b>630","<h3>ID: 46<\/h3><b>Genus_name: <\/b>Ephemerella<br><b>Sequence length: <\/b>658","<h3>ID: 47<\/h3><b>Genus_name: <\/b>Ulmerophlebia<br><b>Sequence length: <\/b>547","<h3>ID: 48<\/h3><b>Species_name: <\/b>Epeorus sylvicola<br><b>Sequence length: <\/b>578","<h3>ID: 49<\/h3><b>Species_name: <\/b>Tricorythodes mosegus<br><b>Sequence length: <\/b>658","<h3>ID: 50<\/h3><b>Species_name: <\/b>Siphlonurus alternatus<br><b>Sequence length: <\/b>658","<h3>ID: 51<\/h3><b>Species_name: <\/b>Leucrocuta hebe<br><b>Sequence length: <\/b>658","<h3>ID: 52<\/h3><b>Species_name: <\/b>Baetis tricaudatus B BG<br><b>Sequence length: <\/b>658","<h3>ID: 53<\/h3><b>Species_name: <\/b>Ephemerella tibialis<br><b>Sequence length: <\/b>658","<h3>ID: 54<\/h3><b>Species_name: <\/b>Paraleptophlebia submarginata<br><b>Sequence length: <\/b>658","<h3>ID: 55<\/h3><b>Species_name: <\/b>Habroleptoides confusa<br><b>Sequence length: <\/b>658","<h3>ID: 56<\/h3><b>Species_name: <\/b>Electrogena ujhelyii<br><b>Sequence length: <\/b>658","<h3>ID: 57<\/h3><b>Species_name: <\/b>Baetodes SADR01<br><b>Sequence length: <\/b>658","<h3>ID: 58<\/h3><b>Genus_name: <\/b>Epeorus<br><b>Sequence length: <\/b>658","<h3>ID: 59<\/h3><b>Species_name: <\/b>Acerpenna macdunnoughi<br><b>Sequence length: <\/b>658","<h3>ID: 60<\/h3><b>Species_name: <\/b>Maccaffertium smithae<br><b>Sequence length: <\/b>658","<h3>ID: 61<\/h3><b>Genus_name: <\/b>Ecdyonurus<br><b>Sequence length: <\/b>636","<h3>ID: 62<\/h3><b>Species_name: <\/b>Baetis tricaudatus<br><b>Sequence length: <\/b>586","<h3>ID: 63<\/h3><b>Genus_name: <\/b>Andesiops<br><b>Sequence length: <\/b>590","<h3>ID: 64<\/h3><b>Species_name: <\/b>Baetis macani<br><b>Sequence length: <\/b>658","<h3>ID: 65<\/h3><b>Genus_name: <\/b>Habroleptoides<br><b>Sequence length: <\/b>658","<h3>ID: 66<\/h3><b>Genus_name: <\/b>Cloeon<br><b>Sequence length: <\/b>658","<h3>ID: 67<\/h3><b>Genus_name: <\/b>Cloeon<br><b>Sequence length: <\/b>658","<h3>ID: 68<\/h3><b>Species_name: <\/b>Ephoron shigae<br><b>Sequence length: <\/b>624","<h3>ID: 69<\/h3><b>Species_name: <\/b>Epeorus assimilis<br><b>Sequence length: <\/b>639","<h3>ID: 70<\/h3><b>Species_name: <\/b>Ephemerella invaria<br><b>Sequence length: <\/b>523","<h3>ID: 71<\/h3><b>Species_name: <\/b>Ephoron album<br><b>Sequence length: <\/b>523","<h3>ID: 72<\/h3><b>Species_name: <\/b>Oligoneuriella rhenana<br><b>Sequence length: <\/b>553","<h3>ID: 73<\/h3><b>Species_name: <\/b>Ephemerella dorothea<br><b>Sequence length: <\/b>639","<h3>ID: 74<\/h3><b>Genus_name: <\/b>Atalophlebia<br><b>Sequence length: <\/b>658","<h3>ID: 75<\/h3><b>Species_name: <\/b>Choroterpes basalis<br><b>Sequence length: <\/b>658","<h3>ID: 76<\/h3><b>Genus_name: <\/b>Cinygmula<br><b>Sequence length: <\/b>658","<h3>ID: 77<\/h3><b>Species_name: <\/b>Dipteromimus tipuliformis<br><b>Sequence length: <\/b>482","<h3>ID: 78<\/h3><b>Genus_name: <\/b>Afrocaenis<br><b>Sequence length: <\/b>658","<h3>ID: 79<\/h3><b>Species_name: <\/b>Baetis tricaudatus<br><b>Sequence length: <\/b>636","<h3>ID: 80<\/h3><b>Genus_name: <\/b>Siphlonurus<br><b>Sequence length: <\/b>658","<h3>ID: 81<\/h3><b>Species_name: <\/b>Maccaffertium vicarium<br><b>Sequence length: <\/b>679","<h3>ID: 82<\/h3><b>Species_name: <\/b>Maccaffertium smithae<br><b>Sequence length: <\/b>658","<h3>ID: 83<\/h3><b>Species_name: <\/b>Paraleptophlebia bicornuta<br><b>Sequence length: <\/b>658","<h3>ID: 84<\/h3><b>Species_name: <\/b>Epeorus longimanus B BG<br><b>Sequence length: <\/b>658","<h3>ID: 85<\/h3><b>Genus_name: <\/b>Prebaetodes<br><b>Sequence length: <\/b>658","<h3>ID: 86<\/h3><b>Genus_name: <\/b>Thraulodes<br><b>Sequence length: <\/b>638","<h3>ID: 87<\/h3><b>Species_name: <\/b>Stenacron interpunctatum<br><b>Sequence length: <\/b>610","<h3>ID: 88<\/h3><b>Species_name: <\/b>Tricorythodes explicatus<br><b>Sequence length: <\/b>658","<h3>ID: 89<\/h3><b>Species_name: <\/b>Acentrella turbida<br><b>Sequence length: <\/b>658","<h3>ID: 90<\/h3><b>Species_name: <\/b>Acerpenna macdunnoughi<br><b>Sequence length: <\/b>658","<h3>ID: 91<\/h3><b>Species_name: <\/b>Leptophlebia cupida<br><b>Sequence length: <\/b>658","<h3>ID: 92<\/h3><b>Species_name: <\/b>Atalophlebia albiterminata<br><b>Sequence length: <\/b>630","<h3>ID: 93<\/h3><b>Species_name: <\/b>Habroleptoides confusa<br><b>Sequence length: <\/b>658","<h3>ID: 94<\/h3><b>Species_name: <\/b>Siphlaenigma janae<br><b>Sequence length: <\/b>658","<h3>ID: 95<\/h3><b>Species_name: <\/b>Baetis vernus<br><b>Sequence length: <\/b>618","<h3>ID: 96<\/h3><b>Species_name: <\/b>Epeorus fragilis<br><b>Sequence length: <\/b>616","<h3>ID: 97<\/h3><b>Species_name: <\/b>Oligoneuriella rhenana<br><b>Sequence length: <\/b>658","<h3>ID: 98<\/h3><b>Species_name: <\/b>Serratella micheneri<br><b>Sequence length: <\/b>658","<h3>ID: 99<\/h3><b>Genus_name: <\/b>Fallceon<br><b>Sequence length: <\/b>658","<h3>ID: 100<\/h3><b>Species_name: <\/b>Heptagenia pulla<br><b>Sequence length: <\/b>658"],null,null,{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null]}],"limits":{"lat":[-42.15,69],"lng":[-122.742,174.51]}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="creating-a-report" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-a-report" class="anchor"></a>Creating a report</h3>
<p>A report is a simple and rapid way to get an overview of the current state of your reference library and to identify some possible issues. You can compile a report using the function <code>refdb_report</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/refdb_report.html">refdb_report</a></span><span class="op">(</span><span class="va">lib</span><span class="op">)</span></code></pre></div>
<p>The result (not shown here) is an interactive HTML report which can be opened in any recent web browser. It contains some statistics and plots, and the results of functions <code>refdb_check_tax_typo</code>, <code>refdb_check_tax_conflict</code> and <code>refdb_check_seq_conflict</code>. These functions can be used to identify possible spelling errors in taxonomic names, conflicts in the taxonomic tree and lack of genetic resolution in sequences, respectively.</p>
</div>
<div id="exporting-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#exporting-the-data" class="anchor"></a>Exporting the data</h3>
<p>Once satisfied with the results, they can be exported for further use. Since the reference database is a dataframe, it is possible to use any export functions (from R or from other packages) that support dataframes. Here we show how to export to CSV using the function <code>write_csv</code> from the <code>readr</code> package.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">lib</span>, <span class="st">"my_reference_library.csv"</span><span class="op">)</span></code></pre></div>
<p>This will write the actual content of <code>lib</code> but ignore the fields which have been previously set. We can save them in a separate YAML file using:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/refdb_write_fields.html">refdb_write_fields</a></span><span class="op">(</span><span class="va">lib</span>, <span class="st">"my_reference_fields.yml"</span><span class="op">)</span></code></pre></div>
<p>Then, the next time you will open this reference database, it will be easy to set the fields again using the argument <code>config_yaml</code> of the function <code>refdb_set_fields</code>:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lib</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"my_reference_library.csv"</span><span class="op">)</span>
<span class="va">lib</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_set_fields.html">refdb_set_fields</a></span><span class="op">(</span><span class="va">lib</span>, config_yaml <span class="op">=</span> <span class="st">"my_reference_fields.yml"</span><span class="op">)</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Francois Keck.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
