<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Downloading and combining data from NCBI Genbank and BOLD • refdb</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Downloading and combining data from NCBI Genbank and BOLD">
<meta property="og:description" content="refdb">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">refdb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro_refdb.html">Introduction to the refdb package</a>
    </li>
    <li>
      <a href="../articles/ncbi_bold.html">Downloading and combining data from NCBI Genbank and BOLD</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fkeck/refdb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ncbi_bold_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Downloading and combining data from NCBI Genbank and BOLD</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/fkeck/refdb/blob/master/vignettes/ncbi_bold.Rmd"><code>vignettes/ncbi_bold.Rmd</code></a></small>
      <div class="hidden name"><code>ncbi_bold.Rmd</code></div>

    </div>

    
    
<p>This short vignette will show you how you can use the functions of the refdb package to download sequence data from two popular repositories: NCBI Genbank and BOLD. We will also cover how these data can be combined into a single reference database. For an introduction to the package refdb see the vignette <em>Introduction to the refdb package</em>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://fkeck.github.io/refdb/">refdb</a></span><span class="op">)</span></code></pre></div>
<div id="getting-data-from-ncbi-genbank" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-data-from-ncbi-genbank" class="anchor"></a>Getting data from NCBI Genbank</h2>
<p>We start by downloading sequence data from NCBI. The refdb packages uses the rentrez package (Winter 2017) to interface with NCBI servers.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>silo_ncbi &lt;-<span class="st"> </span><span class="kw">refdb_import_NCBI</span>(<span class="st">"Silo COI"</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="co">#&gt; Downloading 72 sequences from NCBI...</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co">#&gt; </span></span>
<span id="cb2-4"><a href="#cb2-4"></a>  <span class="op">|</span><span class="st">                                                                                                                                       </span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="st">  </span><span class="er">|</span><span class="st">                                                                                                                                 </span><span class="er">|</span><span class="st">   </span><span class="dv">0</span>%</span>
<span id="cb2-6"><a href="#cb2-6"></a>  <span class="op">|</span><span class="st">                                                                                                                                       </span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="st">  </span><span class="er">|=================================================================================================================================|</span><span class="st"> </span><span class="dv">100</span>%</span></code></pre></div>
<p>Data already have a set of fields defined.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/refdb_get_fields.html">refdb_get_fields</a></span><span class="op">(</span><span class="va">silo_ncbi</span><span class="op">)</span>
<span class="co">#&gt; source: source</span>
<span class="co">#&gt; id: id</span>
<span class="co">#&gt; taxonomy:</span>
<span class="co">#&gt;   superkingdom: superkingdom</span>
<span class="co">#&gt;   kingdom: kingdom</span>
<span class="co">#&gt;   phylum: phylum</span>
<span class="co">#&gt;   subphylum: subphylum</span>
<span class="co">#&gt;   class: class</span>
<span class="co">#&gt;   subclass: subclass</span>
<span class="co">#&gt;   infraclass: infraclass</span>
<span class="co">#&gt;   order: order</span>
<span class="co">#&gt;   suborder: suborder</span>
<span class="co">#&gt;   infraorder: infraorder</span>
<span class="co">#&gt;   superfamily: superfamily</span>
<span class="co">#&gt;   family: family</span>
<span class="co">#&gt;   genus: genus</span>
<span class="co">#&gt;   species: species</span>
<span class="co">#&gt; sequence: sequence</span>
<span class="co">#&gt; marker: gene</span>
<span class="co">#&gt; latitude: latitude</span>
<span class="co">#&gt; longitude: longitude</span></code></pre></div>
</div>
<div id="getting-data-from-bold" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-data-from-bold" class="anchor"></a>Getting data from BOLD</h2>
<p>Now let’s download data for another taxon from the BOLD database.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">goera_bold</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_import_BOLD.html">refdb_import_BOLD</a></span><span class="op">(</span>taxon <span class="op">=</span> <span class="st">"Goera pilosa"</span>, ncbi_taxo <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Downloading 30 sequences from BOLD...</span></code></pre></div>
<p>You may have noticed that the search interface is a bit different. This is because here we rely on another package (bold, Chamberlain 2020) to download the data. You can check the manual of <code>refdb_import_BOLD</code> to see the different arguments available. For the purpose of this vignette we also have turned off the automatic conversion to the NCBI taxonomy.</p>
<p>Similarly to the NCBI data, data downloaded from BOLD have a set of fields defined automatically.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/refdb_get_fields.html">refdb_get_fields</a></span><span class="op">(</span><span class="va">goera_bold</span><span class="op">)</span>
<span class="co">#&gt; source: source</span>
<span class="co">#&gt; id: sequenceID</span>
<span class="co">#&gt; taxonomy:</span>
<span class="co">#&gt;   phylum: phylum_name</span>
<span class="co">#&gt;   class: class_name</span>
<span class="co">#&gt;   order: order_name</span>
<span class="co">#&gt;   family: family_name</span>
<span class="co">#&gt;   subfamily: subfamily_name</span>
<span class="co">#&gt;   genus: genus_name</span>
<span class="co">#&gt;   species: species_name</span>
<span class="co">#&gt; sequence: nucleotides</span>
<span class="co">#&gt; marker: markercode</span>
<span class="co">#&gt; latitude: lat</span>
<span class="co">#&gt; longitude: lon</span></code></pre></div>
</div>
<div id="merging-data-from-different-sources" class="section level2">
<h2 class="hasAnchor">
<a href="#merging-data-from-different-sources" class="anchor"></a>Merging data from different sources</h2>
<p>We can now use <code>refdb_merge</code> to merge the two databases. To make things clearer we will keep only the first three sequences of <code>goera_bold</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Extract the first three records of goera_bold</span>
<span class="va">goera_bold</span> <span class="op">&lt;-</span> <span class="va">goera_bold</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span>

<span class="co"># Merging goera_bold and silo_ncbi into one database</span>
<span class="va">bold_ncbi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_merge.html">refdb_merge</a></span><span class="op">(</span><span class="va">goera_bold</span>, <span class="va">silo_ncbi</span><span class="op">)</span></code></pre></div>
<p>Note that you can merge more than two database with <code>refdb_merge</code>.</p>
<p>Now, let’s take a look at the columns <code>genus_name</code>, <code>species_name</code> and <code>nucleotides</code> of the merged database.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bold_ncbi</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"source"</span>, <span class="st">"genus_name"</span>, <span class="st">"species_name"</span>, <span class="st">"nucleotides"</span><span class="op">)</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 75 x 4</span>
<span class="co">#&gt;    source genus_name species_name  nucleotides                                                                                             </span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;         &lt;DNA&gt;                                                                                                   </span>
<span class="co">#&gt;  1 BOLD   Goera      Goera pilosa  AACAATTTATTTTATTTTTGGTATTTGATCAGGAATAGTCGGAACGTCCCTAAGTATAATTATTCGAATTGAATTAGGAACAGCTAATTCTTTAATTAAAAAT…</span>
<span class="co">#&gt;  2 BOLD   Goera      Goera pilosa  AACAATTTATTTTATTTTTGGTATTTGATCAGGAATAGTCGGAACATCCCTAAGTATAATTATTCGAATTGAATTAGGAACAGCTAATTCTTTAATTAAAAAT…</span>
<span class="co">#&gt;  3 BOLD   Goera      Goera pilosa  AACAATTTATTTTATTTTTGGTATTTGATCAGGAATAGTCGGAACGTCCCTAAGTATAATTATTCGAATTGAATTAGGAACAGCTAATTCTTTAATTAAAAAT…</span>
<span class="co">#&gt;  4 NCBI   Silo       Silo pallipes AACTATTTATTTTATTTTTGGAATTTGATCTGGAATAGTAGGAACTTCTCTAAGAATAATTATTCGAACTGAGCTAGGGACTGCTGAATCTTTAATTAAAAAT…</span>
<span class="co">#&gt;  5 NCBI   Silo       Silo pallipes AACTATTTATTTTATTTTTGGAATTTGATCTGGAATAGTAGGAACTTCTCTAAGAATAATTATTCGAACTGAGCTAGGGACTGCTGAATCTTTAATTAAAAAT…</span>
<span class="co">#&gt;  6 NCBI   Silo       Silo pallipes AACTATTTATTTTATTTTTGGAATTTGATCTGGAATAGTAGGAACTTCTCTAAGAATAATTATTCGAACTGAGCTAGGGACTGCTGAATCTTTAATTAAAAAT…</span>
<span class="co">#&gt;  7 NCBI   Silo       Silo pallipes AACTATTTATTTTATTTTTGGAATTTGATCTGGAATAGTAGGAACTTCTCTAAGAATAATTATTCGAACTGAGCTAGGGACTGCTGAATCTTTAATTAAAAAT…</span>
<span class="co">#&gt;  8 NCBI   Silo       Silo pallipes AACTATTTATTTTATTTTTGGAATTTGATCTGGAATAGTAGGAACTTCTCTAAGAATAATTATTCGAACTGAGCTAGGGACTGCTGAATCTTTAATTAAAAAT…</span>
<span class="co">#&gt;  9 NCBI   Silo       Silo pallipes AACTATTTATTTTATTTTTGGAATTTGATCTGGAATAGTAGGAACTTCTCTAAGAATAATTATTCGAACTGAGCTAGGGACTGCTGAATCTTTAATTAAAAAT…</span>
<span class="co">#&gt; 10 NCBI   Silo       Silo pallipes AACTATTTATTTTATTTTTGGAATTTGATCTGGAATAGTAGGAACTTCTCTAAGAATAATTATTCGAACTGAGCTAGGGACTGCTGAATCTTTAATTAAAAAT…</span>
<span class="co">#&gt; # … with 65 more rows</span></code></pre></div>
<p>Despite those columns having different names in the two databases, they were merged successfully because they were associated to the same field. The names used in the merged database are inherited from the database used as first argument in the merge functions (here <code>goera_bold</code>).</p>
<p>By default only the columns that are associated to a field in one of the two databases are kept. Of course, if there is no equivalent in the other database, NAs are produced. See the columns <code>subfamily_name</code> (subfamily is a taxonomic rank which exist only BOLD):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bold_ncbi</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"source"</span>, <span class="st">"subfamily_name"</span><span class="op">)</span><span class="op">]</span>
<span class="co">#&gt; # A tibble: 75 x 2</span>
<span class="co">#&gt;    source subfamily_name</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;         </span>
<span class="co">#&gt;  1 BOLD   Goerinae      </span>
<span class="co">#&gt;  2 BOLD   Goerinae      </span>
<span class="co">#&gt;  3 BOLD   Goerinae      </span>
<span class="co">#&gt;  4 NCBI   &lt;NA&gt;          </span>
<span class="co">#&gt;  5 NCBI   &lt;NA&gt;          </span>
<span class="co">#&gt;  6 NCBI   &lt;NA&gt;          </span>
<span class="co">#&gt;  7 NCBI   &lt;NA&gt;          </span>
<span class="co">#&gt;  8 NCBI   &lt;NA&gt;          </span>
<span class="co">#&gt;  9 NCBI   &lt;NA&gt;          </span>
<span class="co">#&gt; 10 NCBI   &lt;NA&gt;          </span>
<span class="co">#&gt; # … with 65 more rows</span></code></pre></div>
<p>Alternatively, we can request the function <code>refdb_merge</code> to return only the fields shared by all the reference databases.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bold_ncbi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_merge.html">refdb_merge</a></span><span class="op">(</span><span class="va">goera_bold</span>, <span class="va">silo_ncbi</span>, keep <span class="op">=</span> <span class="st">"fields_shared"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">bold_ncbi</span><span class="op">)</span>
<span class="co">#&gt;  [1] "source"       "sequenceID"   "phylum_name"  "class_name"   "order_name"   "family_name"  "genus_name"   "species_name" "nucleotides" </span>
<span class="co">#&gt; [10] "markercode"   "lat"          "lon"</span></code></pre></div>
<p>NCBI and BOLD share several taxonomic ranks (phylum, class, order, family, genus, and species) that are a good basis for sequence taxonomic classification. But we may also use the default behavior of <code>refdb_import_BOLD</code> to force BOLD data to conform to the NCBI taxonomic system. So let’s download these data again.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">goera_bold</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refdb_import_BOLD.html">refdb_import_BOLD</a></span><span class="op">(</span>taxon <span class="op">=</span> <span class="st">"Goera pilosa"</span>, ncbi_taxo <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Downloading 30 sequences from BOLD...</span>
<span class="co">#&gt; </span>
<span class="va">Processing</span><span class="op">:</span> <span class="va">Goera</span>
<span class="va">goera_bold</span> <span class="op">&lt;-</span> <span class="va">goera_bold</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span>
<span class="fu"><a href="../reference/refdb_get_fields.html">refdb_get_fields</a></span><span class="op">(</span><span class="va">silo_ncbi</span><span class="op">)</span>
<span class="co">#&gt; source: source</span>
<span class="co">#&gt; id: id</span>
<span class="co">#&gt; taxonomy:</span>
<span class="co">#&gt;   superkingdom: superkingdom</span>
<span class="co">#&gt;   kingdom: kingdom</span>
<span class="co">#&gt;   phylum: phylum</span>
<span class="co">#&gt;   subphylum: subphylum</span>
<span class="co">#&gt;   class: class</span>
<span class="co">#&gt;   subclass: subclass</span>
<span class="co">#&gt;   infraclass: infraclass</span>
<span class="co">#&gt;   order: order</span>
<span class="co">#&gt;   suborder: suborder</span>
<span class="co">#&gt;   infraorder: infraorder</span>
<span class="co">#&gt;   superfamily: superfamily</span>
<span class="co">#&gt;   family: family</span>
<span class="co">#&gt;   genus: genus</span>
<span class="co">#&gt;   species: species</span>
<span class="co">#&gt; sequence: sequence</span>
<span class="co">#&gt; marker: gene</span>
<span class="co">#&gt; latitude: latitude</span>
<span class="co">#&gt; longitude: longitude</span>
<span class="fu"><a href="../reference/refdb_get_fields.html">refdb_get_fields</a></span><span class="op">(</span><span class="va">goera_bold</span><span class="op">)</span>
<span class="co">#&gt; source: source</span>
<span class="co">#&gt; id: sequenceID</span>
<span class="co">#&gt; taxonomy:</span>
<span class="co">#&gt;   superkingdom: superkingdom</span>
<span class="co">#&gt;   kingdom: kingdom</span>
<span class="co">#&gt;   phylum: phylum</span>
<span class="co">#&gt;   subphylum: subphylum</span>
<span class="co">#&gt;   class: class</span>
<span class="co">#&gt;   subclass: subclass</span>
<span class="co">#&gt;   infraclass: infraclass</span>
<span class="co">#&gt;   order: order</span>
<span class="co">#&gt;   suborder: suborder</span>
<span class="co">#&gt;   infraorder: infraorder</span>
<span class="co">#&gt;   superfamily: superfamily</span>
<span class="co">#&gt;   family: family</span>
<span class="co">#&gt;   genus: genus</span>
<span class="co">#&gt;   species: species_name</span>
<span class="co">#&gt; sequence: nucleotides</span>
<span class="co">#&gt; marker: markercode</span>
<span class="co">#&gt; latitude: lat</span>
<span class="co">#&gt; longitude: lon</span></code></pre></div>
<p>Now we can observe many more matching taxonomic fields between the two database, which makes the merge operation much more straightforward. Note that you can operate NCBI taxonomic conversion with data from other sources using the function <code>refdb_set_ncbitax</code>.</p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>Winter, D. J. (2017) rentrez: an R package for the NCBI eUtils API The R Journal 9(2):520-526</p>
<p>Chamberlain, S. (2020). bold: Interface to Bold Systems API. R package version 1.1.0. <a href="https://CRAN.R-project.org/package=bold" class="uri">https://CRAN.R-project.org/package=bold</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Francois Keck.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
