<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ refdb</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">refdb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/intro_refdb.html">Introduction to the refdb package</a>
    </li>
    <li>
      <a href="../articles/ncbi_bold.html">Downloading and combining data from NCBI Genbank and BOLD</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/fkeck/refdb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="check_fields.html">check_fields()</a></code> </p>
        </td>
        <td><p>Internal check for fields</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fields_dbs.html">refdb_set_fields_NCBI()</a></code> <code><a href="fields_dbs.html">refdb_set_fields_BOLD()</a></code> <code><a href="fields_dbs.html">refdb_set_fields_PR2()</a></code> <code><a href="fields_dbs.html">refdb_set_fields_diatbarcode()</a></code> </p>
        </td>
        <td><p>Functions to set fields for various databases</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_scores.html">.filter_seq_length()</a></code> </p>
        </td>
        <td><p>Scores for filtering operations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ncbi_taxonomy.html">get_ncbi_taxonomy()</a></code> </p>
        </td>
        <td><p>Get NCBI taxonomy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="igraph_from_taxo.html">igraph_from_taxo()</a></code> </p>
        </td>
        <td><p>Create a graph from a taxonomic table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_ncbi_table.html">make_ncbi_table()</a></code> </p>
        </td>
        <td><p>Parse NCBI XML and make a table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ncbi_taxo_rank.html">ncbi_taxo_rank()</a></code> </p>
        </td>
        <td><p>Taxonomic ranks of the NCBI Taxonomy database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="process_geo_ncbi.html">process_geo_ncbi()</a></code> </p>
        </td>
        <td><p>Process coordinate column returned by NCBI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_check_seq_conflict.html">refdb_check_seq_conflict()</a></code> </p>
        </td>
        <td><p>Check for conflicts in sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_check_tax_conflict.html">refdb_check_tax_conflict()</a></code> </p>
        </td>
        <td><p>Check for conflicts in taxonomy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_check_tax_typo.html">refdb_check_tax_typo()</a></code> </p>
        </td>
        <td><p>Check for typos in taxonomic names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_clean_seq_crop_primers.html">refdb_clean_seq_crop_primers()</a></code> </p>
        </td>
        <td><p>Crop genetic sequences with a set of primers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_clean_seq_remove_gaps.html">refdb_clean_seq_remove_gaps()</a></code> </p>
        </td>
        <td><p>Remove gaps from genetic sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_clean_seq_remove_sideN.html">refdb_clean_seq_remove_sideN()</a></code> </p>
        </td>
        <td><p>Remove repeated side N from genetic sequences</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_clean_tax_NA.html">refdb_clean_tax_NA()</a></code> </p>
        </td>
        <td><p>Convert missing taxonomic names to NA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_clean_tax_harmonize_nomenclature.html">refdb_clean_tax_harmonize_nomenclature()</a></code> </p>
        </td>
        <td><p>Harmonize taxonomic name nomenclature</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_clean_tax_remove_blank.html">refdb_clean_tax_remove_blank()</a></code> </p>
        </td>
        <td><p>Remove blank characters from taxonomic names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_clean_tax_remove_extra.html">refdb_clean_tax_remove_extra()</a></code> </p>
        </td>
        <td><p>Remove extra words from taxonomic names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_clean_tax_remove_subsp.html">refdb_clean_tax_remove_subsp()</a></code> </p>
        </td>
        <td><p>Remove subspecific information from taxonomic names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_clean_tax_remove_uncertainty.html">refdb_clean_tax_remove_uncertainty()</a></code> </p>
        </td>
        <td><p>Remove terms indicating uncertainty in taxonomic names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_export_dada2.html">refdb_export_dada2()</a></code> </p>
        </td>
        <td><p>Export reference database for DADA2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_export_mothur.html">refdb_export_mothur()</a></code> </p>
        </td>
        <td><p>Export reference database for Mothur</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_filter_ref_scope.html">refdb_filter_ref_scope()</a></code> </p>
        </td>
        <td><p>Filter records by taxonomic scope of studies</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_filter_seq_ambiguous.html">refdb_filter_seq_ambiguous()</a></code> </p>
        </td>
        <td><p>Filter sequences based on their number of ambiguous character.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_filter_seq_duplicates.html">refdb_filter_seq_duplicates()</a></code> </p>
        </td>
        <td><p>Filter duplicated sequences.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_filter_seq_homopolymers.html">refdb_filter_seq_homopolymers()</a></code> </p>
        </td>
        <td><p>Filter sequences based on their number of repeated character.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_filter_seq_length.html">refdb_filter_seq_length()</a></code> </p>
        </td>
        <td><p>Filter sequences based on their number of character.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_filter_seq_primer.html">refdb_filter_seq_primer()</a></code> </p>
        </td>
        <td><p>Filter sequences based on the presence of primers.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_filter_seq_stopcodon.html">refdb_filter_seq_stopcodon()</a></code> </p>
        </td>
        <td><p>Filter sequences based on their number of of stop codons.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_filter_tax_precision.html">refdb_filter_tax_precision()</a></code> </p>
        </td>
        <td><p>Filter records based on their taxonomic precision.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_get_fields.html">refdb_get_fields()</a></code> </p>
        </td>
        <td><p>Get fields of a reference database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_import_BOLD.html">refdb_import_BOLD()</a></code> </p>
        </td>
        <td><p>Download and import BOLD records</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_import_NCBI.html">refdb_import_NCBI()</a></code> </p>
        </td>
        <td><p>Download and import NCBI Nucleotide records</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_merge.html">refdb_merge()</a></code> </p>
        </td>
        <td><p>Merge reference databases</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_plot_map.html">refdb_plot_map()</a></code> </p>
        </td>
        <td><p>Plot an interactive map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_plot_seqlen_hist.html">refdb_plot_seqlen_hist()</a></code> </p>
        </td>
        <td><p>Plot an histogram of sequence lengths</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_plot_tax_barplot.html">refdb_plot_tax_barplot()</a></code> </p>
        </td>
        <td><p>Barplots of the number of records for the most represented taxa</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_plot_tax_tree.html">refdb_plot_tax_tree()</a></code> </p>
        </td>
        <td><p>Reference database taxonomy tree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_plot_tax_treemap.html">refdb_plot_tax_treemap()</a></code> </p>
        </td>
        <td><p>Reference database treemap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_report.html">refdb_report()</a></code> </p>
        </td>
        <td><p>Compile a report with different checks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_set_fields.html">refdb_set_fields()</a></code> </p>
        </td>
        <td><p>Associate columns to fields</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_set_ncbitax.html">refdb_set_ncbitax()</a></code> </p>
        </td>
        <td><p>Replace the current taxonomy using the NCBI Taxonomy database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refdb_write_fields.html">refdb_write_fields()</a></code> </p>
        </td>
        <td><p>Write fields to a file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="valid_taxo_rank.html">valid_taxo_rank()</a></code> </p>
        </td>
        <td><p>Ranks considered as valid by refdb</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="xml_extract.html">xml_extract()</a></code> </p>
        </td>
        <td><p>Extract XML elements</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Francois Keck.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


